src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -562,7 +562,22 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
             }
           } else if (reply.canceled()) {
             RAY_LOG(DEBUG) << "Lease canceled " << task_id;
-            RequestNewWorkerIfNeeded(scheduling_key);
+            if (reply.placement_group_removed()) {
+              RAY_LOG(INFO) << "The placement group was removed when leasing a worker, "
+                               "will fail the task.";
+              auto &task_queue = scheduling_key_entry.task_queue;
+              while (!task_queue.empty()) {
+                auto &task_spec = task_queue.front();
+                RAY_UNUSED(task_finisher_->MarkPendingTaskFailed(
+                    task_spec, rpc::ErrorType::TASK_CANCELLED, nullptr));
```

Comment:
the error code `rpc::ErrorType::TASK_CANCELLED` seems ambiguous....  maybe we should add a new type like `rpc::ErrorType::PLACEMENT_GROUP_REMOVED_WHEN_LEASING_WORKER` ?  @rkooo567

Reply:
Yeah good idea. But let’s just use placement_group_removed (when leasing is impl details). We can raise PlacementGroupRemovedError in this way.

Reply:
you mean `rpc::ErrorType::PLACEMENT_GROUP_REMOVED` and throw `PlacementGroupRemovedError ` in python layer?